src/action/network.rs

```diff
@@ -0,0 +1,297 @@
+    network_connection::{Family, Type, State},
+};
+
+use crate::session::{self, Session};
+
+#[derive(Debug)]
+struct Error(netstat2::error::Error);
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.0)
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        write!(fmt, "failed to list network connections: {}", self.0)
+    }
+}
+
+impl From<netstat2::error::Error> for Error {
+
+    fn from(error: netstat2::error::Error) -> Error {
+        Error(error)
+    }
+}
+
+impl Into<session::Error> for Error {
+
+    fn into(self) -> session::Error {
+        session::Error::action(self)
+    }
+}
+
+/// A request type for the network connections action.
+pub struct Request {
+    listening_only: bool,
+}
+
+/// A type that holds brief information about the process.
+struct ProcessInfo {
+    /// Process ID.
+    pid: u32,
+    /// Process name.
+    name: Option<String>,
+}
+
+/// A response type for the network connections action.
+pub struct Response<'a> {
+    /// Information about the socket which is used by connection.
+    socket_info: &'a ProtocolSocketInfo,
+    /// Information about the process that owns the connection.
+    process_info: Option<ProcessInfo>,
+}
+
+/// Gets the IP address type from a given address.
+fn make_family(addr: &IpAddr) -> Family {
+    match addr {
+        IpAddr::V4(_) => Family::Inet,
+        IpAddr::V6(_) => Family::Inet6,
+    }
+}
+
+/// Contructs `NetworkEndpoint` from the specified address and port.
+fn make_network_endpoint(addr: &IpAddr, port: u16) -> NetworkEndpoint {
+    NetworkEndpoint {
+        ip: Some(addr.to_string()),
+        port: Some(port as i32),
+    }
+}
+
+/// Converts [netstat2::TcpState][tcp_state] to `State`
+/// enum used in the protobuf
+///
+/// [tcp_state]: ../../../netstat2/enum.TcpState.html
+fn make_state(state: &TcpState) -> State {
+    match state {
+        TcpState::Unknown => State::Unknown,
+        TcpState::Closed => State::Closed,
+        TcpState::Listen => State::Listen,
+        TcpState::SynSent => State::SynSent,
+        TcpState::SynReceived => State::SynRecv,
+        TcpState::Established => State::Established,
+        TcpState::FinWait1 => State::FinWait1,
+        TcpState::FinWait2 => State::FinWait2,
+        TcpState::CloseWait => State::CloseWait,
+        TcpState::Closing => State::Closing,
+        TcpState::LastAck => State::LastAck,
+        TcpState::TimeWait => State::TimeWait,
+        TcpState::DeleteTcb => State::DeleteTcb,
+    }
+}
+
+/// Creates `NetworkConnection` protobuf message from
+/// [netstat2::ProtocolSocketInfo][protocol_socket_info]. The fields that are
+/// impossible to fill using the socket information are set to `None`.
+///
+/// [protocol_socket_info]: ../../../netstat2/enum.ProtocolSocketInfo.html
+fn make_connection_from_socket_info<'a>(
+    socket_info: &'a ProtocolSocketInfo
```

Comment:
Do you need explicit lifetime annotation? I think lifetime inference should handle this easily.